### PR TITLE
fix: let internal govobj/vote inv request trackers expire after 60 sec

### PR DESCRIPTION
**Pull request by UdjinM6**
## Issue being fixed or feature implemented
Another issue noticed during recent sb voting period. Lots of inv are coming from peers that never send the object/vote they announced. Let's not wait for these forever, 60 seconds should be enough.

## What was done?
Add a "timer" to each request.

## How Has This Been Tested?
Run tests, run a MN on mainnet and check logs.

## Breaking Changes
n/a

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/governance/governance.h [hunk 1]
```diff
@@ -270,8 +270,8 @@ class CGovernanceManager : public GovernanceStore
     int nCachedBlockHeight;
     std::map<uint256, CGovernanceObject> mapPostponedObjects;
     hash_s_t setAdditionalRelayObjects;
-    hash_s_t setRequestedObjects;
-    hash_s_t setRequestedVotes;
+    std::map<uint256, int64_t> mapRequestedObjects;
```
Comment (knst) [inline src/governance/governance.h:273]: if `set` is giving information what that's just a set of requested object, the `map` as a member name prefix doesn't give a context.

Consider renaming mapRequestedObjects to `requestedObjectsWithTime` or `timeOfRequestedObjects` or something else.

Same for `AcceptMessage`: `mapHash` -> `hashesWithTime` ?

File: src/governance/governance.cpp [hunk 2]
```diff
@@ -838,22 +855,25 @@ bool CGovernanceManager::ConfirmInventoryRequest(const CInv& inv)
...
-        setHash = &setRequestedVotes;
+        mapHash = &mapRequestedVotes;
+        set_name = "vote";
         break;
     default:
         return false;
     }
 
-    const auto& [_itr, inserted] = setHash->insert(inv.hash);
+    const auto& [_itr, inserted] = mapHash->emplace(inv.hash, GetTime<std::chrono::seconds>().count() + RELIABLE_PROPAGATION_TIME);
```
Comment (knst) [inline src/governance/governance.cpp:873]: note for myself: `GetTime` return mocked time here

**Review by knst — APPROVED** [review]
utACK 75f4115e4445a3ad71ef14937df6014271a7eaa8

File: src/governance/governance.h (hunk 1 above)
Comment (UdjinM6) [reply src/governance/governance.h]: Good point 👍  Also the logic for these 2 maps is essentially the same so I took it a bit further and squashed them into one, see dea1395.

**Review by knst — DISMISSED** [review]
utACK dea139554a61b9972a34d0bfcfe65c8dfd3099c6

File: src/governance/governance.cpp [hunk 3]
```diff
@@ -837,23 +847,13 @@ bool CGovernanceManager::ConfirmInventoryRequest(const CInv& inv)
...
-        break;
-    case MSG_GOVERNANCE_OBJECT_VOTE:
-        setHash = &setRequestedVotes;
-        break;
-    default:
-        return false;
-    }
-
-    const auto& [_itr, inserted] = setHash->insert(inv.hash);
+    const auto valid_until = GetTime<std::chrono::seconds>().count() + RELIABLE_PROPAGATION_TIME;
```
Comment (PastaPastaPasta) [inline src/governance/governance.cpp:850]: maybe we keep the times here in std::chrono types?

**Comment by coolaj86** [comment]
Could someone please explain simply what this means in terms of the user pipeline?

I created the following proposal with a timestamp backdated to match `start_epoch` (or otherwise one day in the past on the same hour as `start_epoch`):
- <https://rpc.digitalcash.dev/#?method=gobject&params=["get","831672df4230c3578c10ab4c939d912596ca8b9328a93912ddf596699d6a3af9"]&submit>

This was to avoid the following, well-documented dangerous UI bug:
- https://github.com/dashpay/dash/blob/v21.1.x/src/rpc/governance.cpp#L168-L171

In code here:

https://github.com/dashhive/DashGov.js/blob/35fc77ecef083776e81367330c57a98760f1f6cd/dashgov.js#L386

https://github.com/dashhive/DashGov.js/blob/35fc77ecef083776e81367330c57a98760f1f6cd/dashgov.js#L395-L420

I believe what this issue is saying is just that it's going to ignore inventory spam based on real message time from the supposed `gobject submit` that should be sending inventory via p2p and gobject via the mn network, so I believe it won't affect this sort of fix, but I'd like an ACK (or a change to the code to not cement the UI bug further in place at the protocol layer).

**Comment by UdjinM6** [comment]
@coolaj86 this PR has nothing to do with the proposal data itself, it's dealing with an observed misbehaviour on p2p level only. Basically, a peer says "I have a governance object/vote with hash X", we remember this and request it. The object with hash X never arrives but we kept waiting for it forever (not really waiting in networking terms, just remembering that we asked for it earlier). This PR simply lets us forget about such requests after 60 seconds.

**Comment by PastaPastaPasta** [comment]
let's rebase I guess to have CI green? may just be a flaky failure, restarting.

**Review by PastaPastaPasta — APPROVED** [review]
utACK d41d87a5be947fdd980c9fb72613f7f7e440662e; changes to tests for this expiration logic would be appreciated

**Review by knst — APPROVED** [review]
utACK 1f4e1a17edf728f03ef9942e219e354b16ba43de

**Pull request merged**